client-server-server/src/main/scala/zio/raft/server/RaftServer.scala

```diff
@@ -0,0 +1,587 @@
+                  _ <- transport.disconnect(routingId).orDie
+                yield this
+
+          case KeepAlive(timestamp) =>
+            sessions.findSessionByRouting(routingId) match
+              case Some(sessionId) =>
+                for
+                  now <- Clock.instant
+                  _ <- transport.sendMessage(routingId, KeepAliveResponse(timestamp)).orDie
+                  newSessions = sessions.updateExpiry(sessionId, now, config)
+                yield copy(sessions = newSessions)
+              case None =>
+                for
+                  _ <- transport.sendMessage(routingId, SessionClosed(SessionCloseReason.SessionError, None)).orDie
+                  _ <- transport.disconnect(routingId).orDie
+                yield this
+
+          case clientRequest: ClientRequest =>
+            sessions.findSessionByRouting(routingId) match
+              case Some(sessionId) =>
+                for
+                  _ <- raftActionsOut.offer(
+                    RaftAction.ClientRequest(sessionId, clientRequest.requestId, clientRequest.payload)
+                  )
+                yield this
+              case None =>
+                for
+                  _ <- transport.sendMessage(routingId, SessionClosed(SessionCloseReason.SessionError, None)).orDie
+                  _ <- transport.disconnect(routingId).orDie
+                yield this
+
+          case ServerRequestAck(requestId) =>
+            sessions.findSessionByRouting(routingId) match
+              case Some(sessionId) =>
+                for
+                  _ <- raftActionsOut.offer(RaftAction.ServerRequestAck(sessionId, requestId))
+                yield this
+              case None =>
+                for
+                  _ <- transport.sendMessage(routingId, SessionClosed(SessionCloseReason.SessionError, None)).orDie
+                  _ <- transport.disconnect(routingId).orDie
+                yield this
+
+          case CloseSession(reason) =>
+            sessions.findSessionByRouting(routingId) match
+              case Some(sessionId) =>
+                for
+                  _ <- ZIO.logInfo(s"Client closed session $sessionId: $reason")
+                  newSessions = sessions.removeSession(sessionId, routingId)
+                  _ <- raftActionsOut.offer(RaftAction.ExpireSession(sessionId))
+                yield copy(sessions = newSessions)
+              case None =>
+                ZIO.succeed(this)
+
+          case ConnectionClosed =>
+            sessions.findSessionByRouting(routingId) match
+              case Some(sessionId) =>
+                for
+                  _ <- ZIO.logInfo(s"Client connection closed: $sessionId")
+                  newSessions = sessions.disconnect(sessionId, routingId)
+                yield copy(sessions = newSessions)
+              case None =>
+                ZIO.succeed(this)
+    end Leader
+  end ServerState
+
+  /** Immutable sessions management.
+    */
+  case class Sessions(
```

Comment:
I originally thought this will be in the state

Reply:
both, some information here that the state doesn't have:
1. routing ids, those are transient, not part of the state
2. pending sessions, those are waiting to be registered in the state
3. keep alives are managed outside of the state machine for efficiency, it adds some complication, where we need the existing sessions when we step up, to know which session we might need to expire (if no keep-alive)